dashboard/app/controllers/browser_events_controller.rb

```diff
@@ -0,0 +1,69 @@
+# A controller for reporting browser logs and metrics to Cloudwatch
+
+class BrowserEventsController < ApplicationController
+  EXPERIMENT_FLAG_NAME = 'browser-cloudwatch-metrics'
+  FIREHOSE_STUDY_NAME = 'browser-cloudwatch-metrics-errors'
+
+  LOGS_CLIENT = Aws::CloudWatchLogs::Client.new
+  LOG_GROUP_NAME_PREFIX = 'browser-event-logs-'
+
+  before_action :check_preconditions
+
+  # POST /put_logs
+  def put_logs
+    body = JSON.parse(request.body.read)
+
+    return render status :bad_request, json: {message: 'missing required params: logs'} unless body["logs"]
+
+    logs = body["logs"].map do |log_payload|
+      {timestamp: (Time.now.to_f * 1000).to_i, message: decorate_log_payload(log_payload)}
+    end
+
+    resp = LOGS_CLIENT.put_log_events(
+      {
+        log_group_name: "#{LOG_GROUP_NAME_PREFIX}#{Rails.env}",
+        log_stream_name: Rails.env.to_s,
+        log_events: logs
+      }
+    )
+
+    if resp.rejected_log_events_info
+      fallback_log_to_firehose('rejected-log-events', resp.rejected_log_events_info)
+    end
+
+    render status: :ok, json: {}
+  rescue => exception
+    fallback_log_to_firehose('put-log-events-error', {exception: exception, logs: logs})
+    render status: :internal_server_error, json: {error: exception}
+  end
+
+  # Adds shared params to all log objects and convert to stringified JSON
+  private def decorate_log_payload(log_payload)
+    if log_payload.is_a?(String)
+      log_payload = {message: log_payload}
+    end
+
+    log_payload['userId'] = current_user.id if current_user
+    log_payload['signedIn'] = !current_user.nil?
+    log_payload['environment'] = Rails.env.to_s
```

Comment:
Is `Rails.env` the right thing here? @sureshc

Reply:
We'd prefer to use a unique identifier for the specific environment whose web clients are publishing these event logs. The Dashboard fully qualified domain name is a good unique identifier for a given environment. The Rails `environment` actually identifies what type of environment a particular deployment is (`test`, `production`, etc.), and there could be multiple deployments that all use the same environment type. There are multiple adhoc environments, and there are multiple environments where `Rails.env` is `test` (the managed test server `test-studio.code.org`, each continuous integration Drone build, and executing any unit test locally on an engineer's workstation).

Reply:
Got it, so is `CDO.dashboard_hostname` sufficient here then? And when assigning the log stream, can I just use `rack_env`? Seems like that's what `dashboard_hostname` uses internally.